### ISSUE TITLE
[Reproduction] Agent-G² does not recover the reported gains under a canonical fixed-task protocol

### ISSUE THREAD

**Issue by zhangdw156** [ZJU-REAL/Agent-G2#1]
Thank you for releasing Agent-G². I am trying to reproduce the paper on ALFWorld and WebShop and would appreciate help identifying the remaining protocol difference.

## Provenance

I forked this repository at official commit [`a447126`](https://github.com/ZJU-REAL/Agent-G2/commit/a447126b7e67f7dd2ab214a5dfc8a975d21f0eed) into [`zhangdw156/Agent-G2`](https://github.com/zhangdw156/Agent-G2/tree/exp/iclr). My experiment HEAD is [`855092d`](https://github.com/zhangdw156/Agent-G2/commit/855092d515b21772f444138808f5fd2802d36086). The `gmsv/` tree at that HEAD is byte-identical to the official tree; the local changes are environment fairness/evaluation, scale and resource launchers, active-rollout handling, a vLLM sampling-seed fix, and tracking/runtime cleanup. The main fairness change is [`dc57250`](https://github.com/zhangdw156/Agent-G2/commit/dc57250e8994c30656fbe695f3bd2aa27033f345).

## Why I added a fairness layer

The inherited environment path did not define a machine-independent task panel. In the legacy ALFWorld path, filesystem discovery order and independent worker-local environment streams could change task membership/order and could repeat tasks during validation. In WebShop, worker seeds participated in goal and price construction, so the same numeric `goal_idx` could resolve to different task content across workers or machines. Resuming also restarted worker-local validation streams. I identified these behaviors through a local source audit and cross-machine checks, then tracked the fixes in the method fork history.

The fairness layer therefore uses explicit canonical manifests and stable identities: 3,553 ALFWorld training games, all 140 `val_seen` games, all 134 `val_unseen` games, 6,410 WebShop training goals, and 500 WebShop validation goals. It validates counts and uniqueness, uses stable relative ALFWorld paths, constructs WebShop goals with worker-independent local RNG and cross-worker fingerprints, samples WebShop training tasks without replacement before repeating each task for the rollout group, and evaluates every fixed validation task exactly once in deterministic chunks of at most 128. The task batch and rollout group remain 16 × 8.

This changes task assignment and evaluation, so it is not trajectory-equivalent to the paper protocol. Its purpose is to make comparisons across methods and machines use the same underlying tasks. I am not claiming that fairness caused the gap; I am reporting the gap under this controlled protocol.

## Experiment protocol

- Qwen2.5-Instruct at 1.5B, 3B, and 7B; one training seed per cell.
- 150 policy updates; 16 task groups × 8 rollouts; validation every five updates.
- ALFWorld validation: unweighted macro-average of the six family success rates over the fixed 140-task `val_seen` panel.
- WebShop validation: exact success over the fixed 500-task panel.
- Validation uses one stochastic sample per task at temperature 0.4.
- The displayed GRPO control uses the same fairness task pools and evaluator, but it was trained from a separate baseline implementation rather than byte-identical Agent-G² framework code.
- Resource/protocol differences from the paper include four trainer GPUs, ALFWorld prompt length 2,048, horizon 50, and a 150-update budget.

## Results

Paper Table 1 reports Agent-G²/GRPO ALFWorld success of `95.3/72.8` at 1.5B and `98.4/77.6` at 7B, and WebShop exact success of `78.9/56.8` and `84.4/66.1`, respectively. My step-150 fixed-panel endpoints are:

| Scale | ALFWorld Agent-G² / GRPO | WebShop Agent-G² / GRPO |
|---|---:|---:|
| 1.5B | 62.9 / 61.9 | 63.8 / 62.8 |
| 3B | 65.5 / 65.9 | 57.4 / 63.6 |
| 7B | 76.0 / 80.3 | 65.0 / 73.8 |

The 3B row is an extension and has no paper counterpart. These local validation values are not directly interchangeable with the paper's held-out test values, but the method-versus-GRPO pattern is also much smaller than reported and reverses in several cells.

## Paper/repository details I may be interpreting incorrectly

While checking parity, I found several differences between the equations/settings described in the paper and the released code/configuration:

1. The paper forms cluster statistics from task-level success rates `p_i`, whereas the released scheduler appears to pool the eight binary rollout outcomes and take their variance directly.
2. The paper includes prefix-SFT with coefficient 0.5 and reports a large ablation effect, while the released launchers appear to leave the prefix-SFT auxiliary loss disabled.
3. The paper states `K=3`, lambda 1, adaptive sigma, learning rate `1e-5`, cosine decay, and 10% warmup; the public ALFWorld path uses automatic `K=5`, lambda 0.3, fixed sigma 0.25, learning rate `1e-6`, and a constant schedule.
4. The released GRPO normalization operates over turn rows carrying duplicated episode rewards, so longer trajectories receive more mass than a one-advantage-per-trajectory reading of the paper.
5. On WebShop, the expert archive matches about 83.4% of sampled rollouts under the canonical task pool; unmatched task groups still receive GRPO updates but do not update the Gaussian scheduler.

## Questions

1. Is the current public repository/configuration expected to reproduce Table 1, or was a different internal commit/configuration used? Could you share the exact launch commands, task manifests, model revisions, and checkpoint-selection rule?
2. Should the scheduler variance be computed across the per-task `p_i` values, or across pooled binary rollout outcomes as in the released implementation?
3. Should prefix-SFT be active for the main result, and if so, which released option/data path enables the paper's coefficient of 0.5?
4. Which of the paper or repository hyperparameters above are authoritative for reproduction?
5. Does Agent-G² require a task pool fully covered by the expert-prefix archive? If so, how should unmatched WebShop tasks be handled without biasing comparison to GRPO?
6. Are paper numbers final-update results, best-validation checkpoints, or averages selected by another rule, and can the five training seeds be provided?

I can provide the resolved configs, complete scalar CSVs, logs, and per-commit diff audit if useful. Any guidance on the smallest paper-faithful change would be very helpful.

**Comment by zhangdw156** [comment]
<img width="2517" height="2357" alt="Image" src="https://github.com/user-attachments/assets/1bc8a3e9-b198-42cb-9dba-c58f95ecf0ac" />
